### PR TITLE
Feature/backoffice smart filters

### PR DESCRIPTION
**Pull request by tsubik**
Adding more dependent filters by using filter trees. (https://www.pivotaltracker.com/story/show/181286870)

### CODE REVIEW

File: app/admin/operator_document.rb [hunk 1]
```diff
@@ -236,31 +227,28 @@ def scoped_collection
...
-               .arel
-               .on("country_translations.country_id = countries.id and country_translations.locale = '#{I18n.locale}'")
-               .order('required_operator_documents.name')
-           RequiredOperatorDocument.find_by_sql(query.to_sql).map { ["#{_1[:name]} - #{_1[:country_name]}", _1[:id]] }
-         }
-  filter :operator, as: :select,
-                    collection: -> { Operator.order(:name) }
-  filter :fmu, label: I18n.t('activerecord.models.fmu.other'), as: :select,
-               collection: -> { Fmu.with_translations(I18n.locale).order('fmu_translations.name') }
+         collection: -> { RequiredOperatorDocument.with_generic.map { |r| [r.name_with_country, r.id] } }
```
Comment (santostiago) [inline app/admin/operator_document.rb:232]: ❤️

File: app/helpers/hash_helper.rb [hunk 2]
```diff
@@ -1,15 +1,11 @@
...
-    result = {}
-    array.each do |element|
-      if result[element.keys.first].present?
-        result[element.keys.first] << element.values.first unless result[element.keys.first].include?(element.values.first)
-      else
-        result[element.keys.first] = [element.values.first]
-      end
-    end
-    result
+    array.group_by(&:first).transform_values { |v| v.map(&:last) }.delete_if { |k, _v| k.nil? }
```
Comment (santostiago) [inline app/helpers/hash_helper.rb:9]: 👍

File: spec/system_helper.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    process_timeout: 120,
+    inspector: true,
+    headless: ENV["HEADLESS"] != "false"
+  )
+end
+
+# Configure Capybara to use :cuprite driver by default
+Capybara.default_driver = Capybara.javascript_driver = :cuprite
+
+RSpec.configure do |config|
```
Comment (santostiago) [inline spec/system_helper.rb:25]: mmm.. what does this change in the rspec execution?

**Review by santostiago — APPROVED** [review]
Looks great!
But we will lose some dependencies on the observation page. Is that ok?

**Comment by santostiago** [comment]
@tsubik maybe before merging it would be a good idea to squash most of the commits :)

**Comment by tsubik** [comment]
Don't want to squash them as I want to have that other solution (using
dependent ajax select) in git history. There are many commits as I changed
my mind how I will implement this.

File: app/admin/observation.rb [hunk 4]
```diff
@@ -211,6 +198,23 @@ def permitted_params
   filter :updated_at
   filter :deleted_at
 
+  dependent_filters do
+    {
+      country_id: {
+        operator_id: Operator.pluck(:country_id, :id),
+        fmu_id: Fmu.pluck(:country_id, :id),
```
[the remark this answers is not in the document]
Comment (tsubik) [reply app/admin/observation.rb:205]: In this case, that was strictly requested.

File: app/admin/observation_report.rb [hunk 5]
```diff
@@ -33,18 +33,23 @@ def apply_filtering(chain)
...
-  filter :attachment, as: :select
-  filter :user, as: :select, collection: -> { User.order(:name) }
-  filter :observations_country_id_eq,
-         label: I18n.t('activerecord.models.country.one'),
+  filter :observations_country_id,
          as: :select,
-         collection: -> { Country.with_translations(I18n.locale).order('country_translations.name') }
-  filter :observers, label: I18n.t('activerecord.attributes.observation.observers'), as: :select,
-                     collection: -> { Observer.with_translations(I18n.locale).order('observer_translations.name') }
+         label: I18n.t('activerecord.models.country.one'),
```
[the remark this answers is not in the document]
Comment (tsubik) [reply app/admin/observation_report.rb:38]: Yes, I removed all filters Sophie don't want there anymore.

File: lib/active_admin/dependent_filters_extension.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+module ActiveAdmin
+  module DependentFiltersExtension
+    def dependent_filters(&block)
+      # sidebar won't be visible as css should hide it
+      sidebar :dependent_filters, only: :index do
+        filter_tree = block.call
+        filter_tree.each_value { |v| v.transform_values! { |collection| HashHelper.aggregate(collection) } }
+
+        # div class: 'dependent-filters', data: { filters: filter_tree }
```
Comment (tsubik) [inline lib/active_admin/dependent_filters_extension.rb:9]: upps I see I forgot to remove this comment here.

File: spec/system_helper.rb (hunk 3 above)
Comment (tsubik) [reply spec/system_helper.rb:25]: We did not have any system spec before, so we do now. System specs are run along the rest when executing `rspec spec`.

File: app/admin/law.rb [hunk 7]
```diff
@@ -19,16 +19,26 @@ def scoped_collection
...
-  filter :infraction, label: I18n.t('active_admin.laws_page.infraction'), as: :select
-  filter :sanctions, label: I18n.t('active_admin.laws_page.sanctions'), as: :select
+  filter :country,
+         as: :select,
+         collection: -> { Country.joins(:laws).with_translations(I18n.locale).order('country_translations.name') }
+  filter :subcategory,
+         as: :select,
+         collection: -> { Subcategory.joins(:laws).with_translations(I18n.locale).order('subcategory_translations.name') }
+  filter :written_infraction_cont, label: I18n.t('active_admin.laws_page.written_infraction')
+  filter :infraction_cont, label: I18n.t('active_admin.laws_page.infraction')
```
[the remark this answers is not in the document]
Comment (tsubik) [reply app/admin/law.rb:29]: it's ransack `contains` so checking if text contains a string. It's a simple text field without any autocompletion now.

[the remark this answers is not in the document]
Comment (tsubik) [reply app/admin/law.rb:29]: I will ask Sophie again what is more desired, as we can change as you said to look by law id for selected infraction text which will help, but I think they wanted simple text field here.

**Pull request merged**